source/agora/common/TransactionPool.d

```diff
@@ -0,0 +1,267 @@
+
+        // note: can fail. we may want to just recover txes from the network instead.
+        this.db = Database(db_path);
+
+        if (db_exists)
+            logInfo("Loaded database from: %s", db_path);
+
+        // create the table if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS tx_pool " ~
+            "(key BLOB PRIMARY KEY, val BLOB NOT NULL)");
+    }
+
+    /***************************************************************************
+
+        Callback used
+
+        Params:
+            arg = unused (null)
+            code = the error code
+            msg = the error message
+
+    ***************************************************************************/
+
+    private static extern(C) void loggerCallback (void *arg, int code,
+        const(char)* msg) nothrow
+    {
+        try
+        {
+            logError("SQLite error: (%s) %s", code, msg.to!string);
+        }
+        catch (Exception ex)
+        {
+            // should not propagate exceptions to C code
+        }
+    }
+
+    /***************************************************************************
+
+        Shut down the database
+
+        Note: this method must be called explicitly, and not inside of
+        a destructor.
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.db.close();
+    }
+
+    /***************************************************************************
+
+        Add a transaction to the pool
+
+        Params:
+            tx = the transaction to add
+
+        Throws:
+            SqliteException if the transaction failed to be added
+
+    ***************************************************************************/
+
+    public void add (Transaction tx) @safe
+    {
+        auto tx_bytes = serializeFull(tx);
+        () @trusted {
+            db.execute("INSERT INTO tx_pool (key, val) VALUES (?, ?)",
+                hashFull(tx)[], tx_bytes);
+        }();
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the number of transactions in the pool
+
+    ***************************************************************************/
+
+    public size_t length () @safe
+    {
+        // todo: optimize / cache this
+        return () @trusted {
+            return db.execute("SELECT count(*) FROM tx_pool").oneValue!size_t;
+        }();
+    }
+
+    /***************************************************************************
+
+        Params:
+            amount = how many transactions to return
+
+        Returns:
+            an array of 'amount' transactions
+
+        Note:
+            @trusted because most of the body executes non-@safe code,
+            which is assumed to be safe
+
+    ***************************************************************************/
+
+    public Transaction[] take (size_t amount) @trusted
```

Comment:
Urg

Reply:
I'm not sure what to do about this. Almost everything inside calls non-safe code..

Reply:
There is a comment right above it.

Reply:
Right. Okay.